adapters/beachfront/beachfront.go

```diff
@@ -383,103 +470,54 @@ func fallBackDeviceType(request *openrtb.BidRequest) openrtb.DeviceType {
 	return openrtb.DeviceTypeMobileTablet
 }
 
-func getVideoRequests(request *openrtb.BidRequest) ([]beachfrontVideoRequest, []error) {
-	var bfReqs = make([]beachfrontVideoRequest, len(request.Imp))
-	var errs = make([]error, 0, len(request.Imp))
-	var failedRequestIndicies = make([]int, 0)
-
-	for i := 0; i < len(request.Imp); i++ {
-		beachfrontExt, err := getBeachfrontExtension(request.Imp[i])
-
-		if err != nil {
-			// Failed to extract the beachfrontExt, so this request is junk.
-			failedRequestIndicies = append(failedRequestIndicies, i)
-			errs = append(errs, err)
-			continue
-		}
-
-		appid, err := getAppId(beachfrontExt, openrtb_ext.BidTypeVideo)
-
-		if err != nil {
-			// Failed to get an appid, so this request is junk.
-			failedRequestIndicies = append(failedRequestIndicies, i)
-			errs = append(errs, err)
-			continue
-		}
-
-		bfReqs[i].AppId = appid
-
-		if beachfrontExt.VideoResponseType != "" {
-			bfReqs[i].VideoResponseType = beachfrontExt.VideoResponseType
-		} else {
-			bfReqs[i].VideoResponseType = "nurl"
-		}
-
-		bfReqs[i].Request = *request
-		var secure int8
+func prepVideoRequestExt(requestImp openrtb.Imp) (openrtb_ext.ExtImpBeachfront, error) {
+	beachfrontExt, err := getBeachfrontExtension(requestImp)
 
-		if bfReqs[i].Request.Site != nil && bfReqs[i].Request.Site.Domain == "" && bfReqs[i].Request.Site.Page != "" {
-			bfReqs[i].Request.Site.Domain = getDomain(bfReqs[i].Request.Site.Page)
-
-			secure = isSecure(bfReqs[i].Request.Site.Page)
-		}
-
-		if bfReqs[i].Request.App != nil && bfReqs[i].Request.App.Domain == "" && bfReqs[i].Request.App.Bundle != "" {
-			if bfReqs[i].Request.App.Bundle != "" {
-				var chunks = strings.Split(strings.Trim(bfReqs[i].Request.App.Bundle, "_"), ".")
+	if err != nil {
+		return beachfrontExt, err
+	}
 
-				if len(chunks) > 1 {
-					bfReqs[i].Request.App.Domain =
-						fmt.Sprintf("%s.%s", chunks[len(chunks)-(len(chunks)-1)], chunks[0])
-				}
-			}
+	beachfrontExt.AppId, err = getAppId(beachfrontExt, openrtb_ext.BidTypeVideo)
 
-		}
+	if err != nil {
+		return beachfrontExt, err
+	}
 
-		if bfReqs[i].Request.Device.DeviceType == 0 {
-			// More fine graned deviceType methods will be added in the future
-			bfReqs[i].Request.Device.DeviceType = fallBackDeviceType(request)
-		}
+	if beachfrontExt.VideoResponseType != "adm" && beachfrontExt.VideoResponseType != "both" {
+		beachfrontExt.VideoResponseType = "nurl"
+	}
 
-		imp := request.Imp[i]
+	return beachfrontExt, err
+}
 
-		imp.Banner = nil
-		imp.Ext = nil
-		imp.Secure = &secure
+func prepVideoRequest(bfReq videoRequest) videoRequest {
+	if bfReq.Request.Site != nil && bfReq.Request.Site.Domain == "" && bfReq.Request.Site.Page != "" {
+		bfReq.Request.Site.Domain = getDomain(bfReq.Request.Site.Page)
```

Comment:
This could be a race condition. `bfReq.Request` is a shallow copy of the incoming openrtb request. `bfReq.Request.Site` is a pointer to the `Site` object of the original openrtb request, which is shared across all adapters. So this line changes the `Site` object shared across all adapters. The same goes for modifying `App.Domain` below.

Reply:
@hhhjort , As per @mansinahar's suggestion, I ran a few variations with the race detector. I didn't get anything definitive out of doing that directly, but it got me looking into how the validation.sh script runs the race tests. I get now that I was getting the error, same error every time, not because the validation checks were running a race test, but just because it was hitting that race condition in normal validation execution, since the race tests only include a subset of the tests, indicated by the "TestRace" prefix. I've added TestRaceJsonSamples() to beachfront_test.go and added a directory for race tests, which currently just includes a copy of my "all-the-calls" test. It seems that the race was happening about 1/10 times the test was run, though I did at one point run 20 tests in a row with no error. Adding it to the race tests run by './validate.sh --nofmt --cov --race 10' throws the error about 1/2 the time or a little more. 

From what I read on mansinahar's link above, the bit that you point out with the shallow copy is pretty much what they describe as a typical race. So I built a deep copy function. I'm pretty sure that I got that right. **I still get the error**. I'm about to go ahead and push everything, but it will fail the validation. I would appreciate it if you could have a look at my deep copy function. Maybe I'm missing something with that. Even if it is not the cause of this error, what I had before totally looks like a race condition.

Assuming that my deep copy is OK, I'm continuing to hunt this down with a better understanding of what I'm looking for and much more consistently able to reproduce the error.

Reply:
I think I see a problem and I'm attempting a solution which involves making multiple deep copies. I'm not sure if that's OK from a processing expense standpoint, but checking if it at least works locally.

Reply:
Yes, in general deep copies are expensive. Once we get a handle on all the problems, we should convert the deep cop(y/ies) into a semi-deep copy, only copying the pieces we have identified as being subject to change.

Reply:
Locally I have changed getVideoRequests back to taking a pointer to request, then I'm making my deep copies when I create each videoRequest, either nurl or adm to put in the admMap. I think that is more correct, as they aren't then modifying/referencing  each other. I also changed the deepCopyRequest function to only reference requestCopy and it's dereferenced children once it is created, instead of referencing request, which I think improves readability, but nothing else as far as I can tell. 

But I stil get the error on ./validate.sh --nofmt --cov --race 10

Reply:
Yes, we should focus on finding the race condition first, then we can tighten things like this up. Broad, expensive fixes like this can tell us if we are in the right ballpark or not, so they are certainly useful that way. My current suspicion is that having a video imp that has both adm and nurl, or just the combo of video and banner, are somehow colliding each other. They both will make use of the same imp. I just haven't found anywhere where you are writing to a shared imp that would cause the issue to appear. The search has uncovered a few nitpicks as you have seen. :)

Reply:
The critical thing that got these changes going on my end is that if an adm request is made to the adapter with multiple impressions, beachfront wants a single request to their backend with all of those impressions, instead of a request for each impression as in the current master. A lot of this trying to mix things up with different types of requests is not really part of their requirement, but seemed like added functionality that could be included. Maybe it can't. If that's the case, maybe I need to take a step back and establish some rules for what requests will be accepted. As far as real life situations, this test is really an extreme edge case. I am up for a challenge though, and not giving up just yet.

Since I created the TestRaceJsonSamples() I'll come up with several tests that hammer it, but not _all_ of the possible combinations at once and see what happens with that.

Thank you for taking the time to nitpick.